File: blue/src/colors_compute/registration.py
"""Read-only registration collision checks. No imports, deletions or retries."""
import asyncio
import inspect
import json
import os
from importlib.resources import files
from urllib.parse import urlsplit, parse_qsl, urlencode
from urllib.request import Request, build_opener, HTTPRedirectHandler, ProxyHandler

from .contract import _missing, _safe


class _NoRedirect(HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


def _get(url, headers):
    with build_opener(ProxyHandler({}), _NoRedirect()).open(Request(url, headers=headers), timeout=30) as response:
        if response.status != 200:
            raise ValueError()
        body = response.read(2097153)
        if len(body) > 2097152:
            raise ValueError()
        return body


async def _http(url, headers):
    return await asyncio.to_thread(_get, url, headers)


def _identifier(value):
    if isinstance(value, str) and not _missing(value):
        return value
    if type(value) in (int, float) and 0 < value <= 9007199254740991 and value == int(value):
        return str(int(value))
    raise ValueError()


def _public(value):
    if not isinstance(value, str) or len(value.split()) < 2:
        raise ValueError()
    return ' '.join(value.split()[:2])


async def registration_preflight(opts, key_mode, ownership=None, public_key=None, environment=None, http=None):
    """Return checked/skipped; failures raise only fixed, credential-free text."""
    try:
        descriptors = json.loads(files('colors_compute').joinpath('registration-preflight.json').read_text())
        provider = opts.get('provider-compute')
        if key_mode not in ('managed', 'external') or not _safe(opts.get('profile')):
            raise ValueError()
        if key_mode == 'external' or provider not in descriptors or opts.get('blue/event') == 'build' or opts.get('blue/dry-run') is True:
            return {'status': 'skipped'}
        descriptor = descriptors[provider]
        if ownership is not None:
            if not isinstance(ownership, dict) or set(ownership) != {'provider', 'scope', 'id'} or ownership['provider'] != provider or ownership['scope'] != descriptor['scope']:
                raise ValueError()
            owned_id = _identifier(ownership['id'])
        else:
            owned_id = None
        material = _public(public_key) if public_key is not None else None
        source = os.environ if environment is None else environment
        token = source.get(descriptor['token'])
        if not isinstance(token, str) or _missing(token) or '\n' in token or '\r' in token:
            raise ValueError()
        headers = {'Authorization': 'Bearer ' + token, 'Accept': 'application/json'}
        endpoint = descriptor['endpoint']
        first = urlsplit(endpoint)
        query = {'per_page': descriptor['per_page']}
        if descriptor['pagination'] == 'page':
            query['page'] = 1
        url = endpoint + '?' + urlencode(query)
        visited, identifiers, matching, owned_seen = set(), set(), [], False
        for _ in range(1000):
            parsed = urlsplit(url)
            pairs = parse_qsl(parsed.query, keep_blank_values=True, strict_parsing=True)
            allowed = {'per_page', 'page'} if descriptor['pagination'] == 'url' else {'per_page', 'page' if descriptor['pagination'] == 'page' else 'cursor'}
            if (len(url) > 4096 or parsed.scheme != 'https' or parsed.netloc != first.netloc or parsed.path != first.path or parsed.fragment
                    or parsed.username or parsed.password or len(dict(pairs)) != len(pairs) or set(dict(pairs)) - allowed or url in visited):
                raise ValueError()
            visited.add(url)
            body = (http or _http)(url, headers.copy())
            if inspect.isawaitable(body):
                body = await body
            if not isinstance(body, bytes) or len(body) > 2097152:
                raise ValueError()
            data = json.loads(body.decode('utf-8'), parse_constant=lambda _: (_ for _ in ()).throw(ValueError()))
            if not isinstance(data, dict) or not isinstance(data.get('ssh_keys'), list):
                raise ValueError()
            for entry in data['ssh_keys']:
                if not isinstance(entry, dict) or not isinstance(entry.get('name'), str):
                    raise ValueError()
                identity = _identifier(entry.get('id'))
                public = _public(entry.get(descriptor['public']))
                if identity in identifiers or len(identifiers) >= 100000:
                    raise ValueError()
                identifiers.add(identity)
                if identity == owned_id:
                    owned_seen = True
                if entry['name'] == opts['profile']:
                    matching.append((identity, public))
            kind = descriptor['pagination']
            if kind == 'url':
                links = data.get('links', {})
                if not isinstance(links, dict) or not isinstance(links.get('pages', {}), dict):
                    raise ValueError()
                next_value = links.get('pages', {}).get('next')
            else:
                meta = data.get('meta')
                section = meta.get('pagination' if kind == 'page' else 'links') if isinstance(meta, dict) else None
                field = 'next_page' if kind == 'page' else 'next'
                if not isinstance(section, dict) or field not in section:
                    raise ValueError()
                next_value = section[field]
            if next_value is None or kind == 'cursor' and next_value == '':
                break
            if kind == 'page':
                if type(next_value) is not int or next_value < 1:
                    raise ValueError()
                url = endpoint + '?' + urlencode({'per_page': descriptor['per_page'], 'page': next_value})
            else:
                if not isinstance(next_value, str) or not next_value:
                    raise ValueError()
                url = next_value if kind == 'url' else endpoint + '?' + urlencode({'per_page': descriptor['per_page'], 'cursor': next_value})
        else:
            raise ValueError()
        if owned_id is not None and not owned_seen:
            raise ValueError()
        for identity, public in matching:
            if identity != owned_id:
                if material is not None and material == public:
                    raise RuntimeError('unowned SSH registration; verify whether hosts survive before explicit recovery')
                raise RuntimeError('foreign SSH registration; do not delete it; investigate or change profile')
            if material is not None and material != public:
                raise ValueError()
        return {'status': 'checked'}
    except RuntimeError as error:
        if str(error) in ('unowned SSH registration; verify whether hosts survive before explicit recovery', 'foreign SSH registration; do not delete it; investigate or change profile'):
            raise ValueError(str(error)) from None
        raise ValueError('SSH registration preflight failed') from None
    except Exception:
        raise ValueError('SSH registration preflight failed') from None

File: red/src/registration.ts
import {get} from 'node:https';
import descriptors from '../resources/registration-preflight.json';

type Map = Record<string, any>;
type HTTP = (url: string, headers: Record<string, string>) => Promise<Uint8Array> | Uint8Array;
const missing = (v: any) => v == null || typeof v === 'string' && (!v.trim() || v.trim().toUpperCase() === 'REPLACE_ME');
const errors = new Set(['unowned SSH registration; verify whether hosts survive before explicit recovery',
  'foreign SSH registration; do not delete it; investigate or change profile']);

const nativeHTTP: HTTP = (url, headers) => new Promise((resolve, reject) => {
  let request: ReturnType<typeof get>;
  const timer = setTimeout(() => request.destroy(Error('request timeout')), 30000);
  request = get(url, {headers}, response => {
    if (response.statusCode !== 200) { response.destroy(); request.destroy(Error('request failed')); return; }
    const chunks: Buffer[] = [];
    let size = 0;
    response.on('data', chunk => {
      size += chunk.length;
      if (size > 2097152) request.destroy(Error('response too large'));
      else chunks.push(Buffer.from(chunk));
    });
    response.on('error', reject);
    response.on('end', () => { clearTimeout(timer); resolve(Buffer.concat(chunks)); });
  });
  request.on('error', error => { clearTimeout(timer); reject(error); });
});

function identifier(value: any): string {
  if (typeof value === 'string' && !missing(value)) return value;
  if (typeof value === 'number' && Number.isSafeInteger(value) && value > 0) return String(value);
  throw Error();
}
function material(value: any): string {
  if (typeof value !== 'string' || value.trim().split(/\s+/).length < 2) throw Error();
  return value.trim().split(/\s+/).slice(0, 2).join(' ');
}
const object = (value: any): value is Map => !!value && typeof value === 'object' && !Array.isArray(value);

export async function registrationPreflight(opts: Map, keyMode: string, ownership: Map | null = null,
  publicKey: string | null = null, environment: Record<string, string | undefined> | null = null, http: HTTP = nativeHTTP) {
  try {
    const provider = opts['provider-compute'];
    if (!['managed', 'external'].includes(keyMode) || typeof opts.profile !== 'string' || !/^[a-zA-Z0-9][a-zA-Z0-9_-]{0,62}$/.test(opts.profile)) throw Error();
    if (keyMode === 'external' || !Object.hasOwn(descriptors, provider) || opts['red/event'] === 'build' || opts['red/dry-run'] === true) return {status: 'skipped'};
    const descriptor = (descriptors as Map)[provider];
    let owned: string | null = null;
    if (ownership !== null) {
      if (!object(ownership) || Object.keys(ownership).sort().join(',') !== 'id,provider,scope' || ownership.provider !== provider || ownership.scope !== descriptor.scope) throw Error();
      owned = identifier(ownership.id);
    }
    const publicMaterial = publicKey === null ? null : material(publicKey);
    const token = (environment ?? process.env)[descriptor.token];
    if (typeof token !== 'string' || missing(token) || /[\r\n]/.test(token)) throw Error();
    const headers = {Authorization: 'Bearer ' + token, Accept: 'application/json'};
    const endpoint = new URL(descriptor.endpoint);
    const query = new URLSearchParams({per_page: String(descriptor.per_page)});
    if (descriptor.pagination === 'page') query.set('page', '1');
    let url = descriptor.endpoint + '?' + query;
    const visited = new Set<string>(), ids = new Set<string>();
    const matching: [string, string][] = [];
    let ownedSeen = false, ended = false;
    for (let page = 0; page < 1000; page++) {
      const parsed = new URL(url);
      const pairs = [...parsed.searchParams];
      const allowed = descriptor.pagination === 'cursor' ? ['per_page', 'cursor'] : ['per_page', 'page'];
      if (url.length > 4096 || parsed.protocol !== 'https:' || parsed.host !== endpoint.host || parsed.pathname !== endpoint.pathname || parsed.hash || parsed.username || parsed.password || new Set(pairs.map(([key]) => key)).size !== pairs.length || pairs.some(([key]) => !allowed.includes(key)) || visited.has(url)) throw Error();
      visited.add(url);
      const body = await http(url, {...headers});
      if (!(body instanceof Uint8Array) || body.length > 2097152) throw Error();
      const data = JSON.parse(new TextDecoder('utf-8', {fatal: true, ignoreBOM: true}).decode(body));
      if (!object(data) || !Array.isArray(data.ssh_keys)) throw Error();
      for (const entry of data.ssh_keys) {
        if (!object(entry) || typeof entry.name !== 'string') throw Error();
        const id = identifier(entry.id), publicValue = material(entry[descriptor.public]);
        if (ids.has(id) || ids.size >= 100000) throw Error();
        ids.add(id);
        if (id === owned) ownedSeen = true;
        if (entry.name === opts.profile) matching.push([id, publicValue]);
      }
      const kind = descriptor.pagination;
      let next: any;
      if (kind === 'url') {
        const links = data.links ?? {};
        if (!object(links) || !object(links.pages ?? {})) throw Error();
        next = links.pages?.next;
      } else {
        const section = data.meta?.[kind === 'page' ? 'pagination' : 'links'];
        const field = kind === 'page' ? 'next_page' : 'next';
        if (!object(section) || !Object.hasOwn(section, field)) throw Error();
        next = section[field];
      }
      if (next == null || kind === 'cursor' && next === '') { ended = true; break; }
      if (kind === 'page') {
        if (typeof next !== 'number' || !Number.isSafeInteger(next) || next < 1) throw Error();
        url = descriptor.endpoint + '?' + new URLSearchParams({per_page: String(descriptor.per_page), page: String(next)});
      } else {
        if (typeof next !== 'string' || !next) throw Error();
        url = kind === 'url' ? next : descriptor.endpoint + '?' + new URLSearchParams({per_page: String(descriptor.per_page), cursor: next});
      }
    }
    if (!ended || owned !== null && !ownedSeen) throw Error();
    for (const [id, publicValue] of matching) {
      if (id !== owned) throw Error(publicMaterial !== null && publicMaterial === publicValue
        ? 'unowned SSH registration; verify whether hosts survive before explicit recovery'
        : 'foreign SSH registration; do not delete it; investigate or change profile');
      if (publicMaterial !== null && publicMaterial !== publicValue) throw Error();
    }
    return {status: 'checked'};
  } catch (error) {
    throw Error(error instanceof Error && errors.has(error.message) ? error.message : 'SSH registration preflight failed');
  }
}
